### PR TITLE
Feature template dates

### PR DESCRIPTION
**Pull request by Liquidmantis**
This adds multiple new template tokens for relative calendar offsets.  These are useful for creating templatized links in daily and weekly notes for navigating to notes along a timeline.  Additionally, it facilitates creating a weekly note template that can have links to the days for that week.

- prevday
- nextday
- prevweek
- nextweek
- isoweek (for consistency with prevweek and next week)
- days of the week by name

### CODE REVIEW

File: lua/telekasten.lua [hunk 1]
```diff
@@ -174,13 +174,13 @@ local function daysuffix(day)
 end
 
 local daymap = {
+    "Sunday",
```
Comment (renerocksai) [inline lua/telekasten.lua:177]: Have you checked if this breaks `daymap[weekday]` in `CalendarAction()`? It looks like it might.

File: lua/telekasten.lua [hunk 2]
```diff
@@ -197,17 +197,23 @@ local monthmap = {
...
-    opts.date = os.date("%Y-%m-%d")
-    local wday = dinfo.wday - 1
-    if wday == 0 then
-        wday = 7
-    end
-    if wday == 6 then
-        wday = 1
-    end
+    opts.date = os.date(df.date)
+    local wday = dinfo.wday - 1 -- compensate for 1-indexed os.date output
```
Comment (renerocksai) [inline lua/telekasten.lua:216]: I haven't tested it yet. But in lua, everything is 1-indexed. Could this ever become 0 - and shouldn't it then wrap over to 7?

File: lua/telekasten.lua [hunk 3]
```diff
@@ -216,20 +222,73 @@ local function calenderinfo_today()
...
+    -- things get a bit hairy at the year rollover.  W01 only starts the first week ofs
+    -- January if it has more than 3 days. Partial weeks with less than 4 days are 
+    -- considered W52, but os.date still sets the year as the new year, so Jan 1 2022 
+    -- would appear as being in 2022-W52.  That breaks linear linking respective 
+    -- of next/prev week, so we want to put the days of that partial week in
+    -- January in 2021-W52.  This tweak will only change the ISO formatted week string.
+    if opts.week == 52 and opts.month == 1 then
+      opts.isoweek = os.date(df.isoweek, time - oneyear) 
+    end
+
```
Comment (renerocksai) [inline lua/telekasten.lua:247]: I agree with that fix!

File: lua/telekasten.lua [hunk 4]
```diff
@@ -216,20 +222,73 @@ local function calenderinfo_today()
...
+    opts.friday    = os.date(df.date, starting_sunday + (5 * oneday))
+    opts.saturday  = os.date(df.date, starting_sunday + (6 * oneday))
+    opts.sunday    = os.date(df.date, starting_sunday + (sunday_offset * oneday))
+
     return opts
 end
 
 local function linesubst(line, title, calendar_info)
-    local cinfo = calendar_info or calenderinfo_today()
+    local dates = calculate_dates(calendar_info)
```
Comment (renerocksai) [inline lua/telekasten.lua:269]: while this is correct ... :smile:

File: lua/telekasten.lua [hunk 5]
```diff
@@ -1082,7 +1141,7 @@ end
 -- find today's daily note and create it if necessary.
 --
 local function GotoToday(opts)
-    opts = opts or calenderinfo_today()
+    opts = opts or calculate_dates()
```
Comment (renerocksai) [inline lua/telekasten.lua:1144]: ... and this also seems correct .... there is one omission which I'll write in the overall review text

**Review by renerocksai — CHANGES_REQUESTED** [review]
Really good work you have done here.

Some minor nitpick I didn't want to bring up earlier. But if it's not asked too much: can you consider trying `stylua`? Your editor puts TABs everywhere which messes with a) non-whitespace-aware diffs and b) catches me unpreparedly when I try to paste code snippets here and in the preview my additions are indented wrongly :smile: 


I use `coc-stylua` so I can `:Format` but there are other options with LSP as well. You actually don't need vim integration, it works as a standalone program, too. You can call it with no options (they are in `.stylua.toml`). But that wouldn't stop me from merging.

Please integrate a call to `calculate_dates(cinfo)` into `CalendarAction()` or else it will never be called when you pick a date from the calendar: 

```lua
local function GotoToday(opts)
    opts = opts or calculate_dates()
        -- if opts is not nil (because CalendarAction sets it) - there will be nothing calculated.
        -- but CalendarAction() does not know about relative dates
        -- because there is no call to calculate_dates(cinfo)
        -- ...
end
```

The more I think about it, maybe it is fixed by passing `opts` like this

```lua
local function GotoToday(opts)
    opts = calculate_dates(opts)
        -- if opts is not nil (because CalendarAction sets it) - the relative dates are calculated
        -- if opts is nil, caclulate_days() will set everything based on today's date 
        -- ...
end
```
 

There is one more caveat: what if someone passes in some non-calendar-related opts you don't know about? Maybe for styling the picker? You erase them with a call to `calculate_dates()`.

My initial suggestion was to just add to the passed-in `cinfo` from within `calculate_dates(cinfo)`, like so:

```lua
local function add_some_stuff_to(cinfo)
    cinfo.some_stuff = 'stuff'
end
```

That way you don't need to return a locally-generated opts - which could potentially erase any other non-date-related opts, like `opts.calendar = true` from `CalendarAction()` -- which is used further down in `GotoToday()`:

```lua
local function GotoToday(opts)
    opts = opts or calculate_dates()

    -- ...

    find_files_sorted({

        -- ...

        attach_mappings = function(prompt_bufnr, map)
            actions.select_default:replace(function()
                actions.close(prompt_bufnr)

                -- open the new note
                if opts.calendar == true then
                -- ^^^^^^^^^^^^^^^^^^^^^ ............. <<<<<<<<<< HERE
                    vim.cmd("wincmd w")
                end
    -- ...
```

File: lua/telekasten.lua (hunk 2 above)
Comment (Liquidmantis) [reply lua/telekasten.lua]: Yeah,  I've been banging my head around this for a while and ended up reverting some of the changes, including the reindexing above.  I had to find the just-right magic combination for the days of week to work for all days, with Sunday considered the last day of the week.

I also ended up finding out that all of lua is 1-indexed as part of this "exercise", so... yay?

Comment (renerocksai) [reply lua/telekasten.lua]: I saw your push and immediately marked this part as resolved :smile: 

Please test it properly for all sorts of edge cases. The calendar plugin can be a great tool for that. (hint!)

**Comment by Liquidmantis** [comment]
Gross!  I didn't realize it was adding tabs.  That's really strange because I run 2 spaces globally.  Must be something with a lua extension overriding it.  

Well, that's really weird.  I thought to turn on my whitespacing and I'm all spaces.  I don't know where the tabs would be coming from.

File: lua/telekasten.lua (hunk 5 above)
Comment (Liquidmantis) [reply lua/telekasten.lua:1151]: Let me keep reading through this, but so far I'm not seeing in.  In this case all I did was replace calendarinfo_today with calculate_dates, but I didn't change the signature.  

I do get now how if opts is not nil that the dates won't be calculated.  I think I was getting confused because `calendarinfo_today` had a local table named `opts` that I thought was your "outputs".  I didn't realize you're passing options around. 🤣

File: lua/telekasten.lua (hunk 2 above)
Comment (Liquidmantis) [reply lua/telekasten.lua]: @renerocksai  Yup!  I've been using the calendar all afternoon for testing!  I found some issues around the calendar that I want to submit PRs for.  Mostly some documentation around using the calendar filetype to make sure some UI  chrome gets disabled.  My signcolumn, relativenumbers, and alignment markers were causing display issues.

It seems like calendar_monday doesn't get respected either.  No matter what the calendar starts the display on Monday.  Maybe I don't understand that setting.

**Comment by renerocksai** [comment]
> Gross! I didn't realize it was adding tabs. That's really strange because I run 2 spaces globally. Must be something with a lua extension overriding it.
> 
> Well, that's really weird. I thought to turn on my whitespacing and I'm all spaces. I don't know where the tabs would be coming from.
> 
> I did get stylua wired up, so let me see if that changes it. When I look at the raw version in my repo it's showing four spaces for indent (locally my autoindent has been aligning so my 2-spaces doesn't change the alignment.)

I might have accused you of TAB-bing wrongly! Apologies! I had noticed my LSP complaining about trailing spaces and tried out cmd-line luastyle to verify that it works from there, too. That might have introduced them here, locally.

**Comment by renerocksai** [comment]
For some reason, I don't see this msg in the thread of the pull request,
but only in my mails.

The assumption that you didn't change the signature and therefore
everything should work is invalid because you changed the semantics and
state of "the program". Previously, the data in the passed-in opts was all
that was needed. If nothing was passed in, it was easily generated. Now,
the data that gets passed in is a) NOT sufficient anymore. But it is b)
important nevertheless. Because of a) you need to extend instead of
overwriting it. Because of b) you need to pass it to your function or else
your function will default to 'today'.

Think of it as everyone adding their bit to the 'opts'. The caller might
add picker-specific or calendar-specific options, you add relative date
specific options, etc. Having opts flying around everywhere and being
captured in lambdas / picker callback functions seems to be a
telescope-ism.

**Comment by Liquidmantis** [comment]
Yeah, sorry, that was my fault.  I made some replies but then as I got my head wrapped more around various bits I deleted them.  I'm making what might be some controversial changes now, where I'm decoupling the data from the options.  I'm trying to figure out an issue with the CalendarAction call but I'm interested to see if you'll find these copacetic or not.  Basically rather than extending opts, especially since lua doesn't have built-ins for table concatenation, I'm passing them around side-by-side.  For some reason I can't see it's not taking the date from CalendarAction now. :/

**Comment by Liquidmantis** [comment]
Alright!  Time for me to call it a day!  I'm removing the DRAFT tag, as I'm satisfied with this submission.  I went a bit heavier handed on this pass, but I think it simplified some things and got rid of a little bit of duplicated code.  As always I'm open to all comments so let me know if any of this grates on your design.

**Review by renerocksai — APPROVED** [review]
WOW!!!! I have to say: What a progression! 
⭐⭐⭐⭐⭐ (verified review) :grin: 

This looks **really** good now! I like how you separated GotoToday and GotoDate. We will do the same with GotoThisWeek once we can click on weeks in the calendar!

I only tested it minimally, checking if it would do the right thing with `{{prevday}}` from a CalendarAction and that worked like a charm.

I also got an idea: now with all those relative dates handy, we could use a command that re-runs the template magic over the current note buffer. It's super handy to write: 'did this on {{monday}}' - and expand the monday. I would like that. Of course, I would not try to be super-fancy to detect daily notes and use their dates as reference date. The moment you type `{{monday}}` and have it converted, is the reference date.

But I digress. I will approve this now!

Many thanks for putting in all this effort and putting up with my picky comments!

👍

**Comment by renerocksai** [comment]
I only added 1 commit on top of it after having run stylua and luacheck, so the GitHub actions will be happy.

Great work!

**Comment by renerocksai** [comment]
Enjoy your feature-improved telekasten :grin: !
